### PR TITLE
Fix odata discriminator

### PR DESCRIPTION
**Pull request by sarangan12**
This PR is to fix the 2 issues reported to:

1. https://github.com/Azure/autorest.typescript/issues/802
2. https://github.com/Azure/autorest.typescript/issues/803
3. https://github.com/Azure/autorest.typescript/issues/805

### CODE REVIEW

File: test/integration/generated/bodyComplex/src/models/index.ts [hunk 1]
```diff
@@ -8,11 +8,11 @@
 
 import * as coreHttp from "@azure/core-http";
 
-export type FishUnion = SalmonUnion | SharkUnion;
-export type DotFishUnion = DotSalmon;
-export type MyBaseTypeUnion = MyDerivedType;
-export type SalmonUnion = SmartSalmon;
-export type SharkUnion = Sawshark | Goblinshark | Cookiecuttershark;
+export type FishUnion = Fish | SalmonUnion | SharkUnion;
```
Comment (joheredi) [inline test/integration/generated/bodyComplex/src/models/index.ts:11]: Why is fish being added here?

File: src/generators/modelsGenerator.ts [hunk 2]
```diff
@@ -536,9 +536,7 @@ function writeUniontypes({ objects }: ClientDetails, modelsFile: SourceFile) {
     .forEach(obj => {
       const polymorphicObject = obj as PolymorphicObjectDetails;
       const childrenNames = [
-        ...(isPolymorphicParentInUnion(polymorphicObject)
```
Comment (joheredi) [inline src/generators/modelsGenerator.ts:539]: I don't think we have to change this

Comment (sarangan12) [reply src/generators/modelsGenerator.ts:539]: We have to change this. I found this after our call. If you look into this file https://github.com/Azure/autorest.typescript/pull/804/files#diff-51f86c8fa85803f85a9039f559710dc4b574b28f7db2ca55e51743d26cd6dd37L509, the build will fail if you do not add the base class to the union.

Comment (sarangan12) [reply src/generators/modelsGenerator.ts:539]: @joheredi without this change, you will see build errors such as:

```
Type 'Shark' is not assignable to type 'FishUnion'.
    Type 'Shark' is not assignable to type 'Cookiecuttershark'.
      Type 'Shark' is not assignable to type '{ fishtype: "cookiecuttershark"; }'.
        Types of property 'fishtype' are incompatible.
          Type '"shark"' is not assignable to type '"cookiecuttershark"'.
```

File: test/integration/generated/bodyComplex/src/models/index.ts (hunk 1 above)
Comment (sarangan12) [reply test/integration/generated/bodyComplex/src/models/index.ts:11]: Discussed Offline. Resolving

File: src/generators/modelsGenerator.ts (hunk 2 above)
Comment (sarangan12) [reply src/generators/modelsGenerator.ts:539]: Discussed Offline. Resolving

**Review by joheredi — APPROVED** [review]
LGTM

**Pull request merged**